packages/iterator-utilities/src/lib/equalBy.ts

```diff
@@ -0,0 +1,44 @@
+import { from, type IterableResolvable } from './from';
+import { assertFunction } from './shared/_assertFunction';
+import { toIterableIterator } from './toIterableIterator';
+
+/**
+ * Determines if the elements of both iterators are equal with respect to the specified equality function.
+ *
+ * @param iterable The iterator to compare.
+ * @param other The iterator to compare against.
+ * @returns Whether the two iterators are equal with respect to the specified equality function.
+ *
+ * @example
+ * ```typescript
+ * import { equalBy } from '@sapphire/iterator-utilities';
+ *
+ * const x = [1, 2, 3, 4];
+ * const y = [1, 4, 9, 16];
+ *
+ * console.log(equalBy(x, y, (a, b) => a * a === b));
+ * // Output: true
+ * ```
+ *
+ * @remarks
+ *
+ * This function consumes the entire iterator.
+ */
+export function equalBy<const ElementType>(
+	iterable: IterableResolvable<ElementType>,
+	other: IterableResolvable<ElementType>,
+	callbackFn: (x: ElementType, y: ElementType) => boolean
+): boolean {
+	callbackFn = assertFunction(callbackFn);
+
+	const iterator1 = from(other);
+
+	for (const value0 of toIterableIterator(iterable)) {
+		const result1 = iterator1.next();
+		if (result1.done || !callbackFn(value0, result1.value)) return false;
```

Comment:
Can't we get a fastpath for arrays?

Reply:
Quite a lot of things could get a fastpath for arrays, but that'll require adding more code to handle those cases.

That being said, I considered the possibility of having a `sizeHint` in our iterator.